CMakeLists.txt

```diff
@@ -257,19 +256,27 @@ option(INSTALL_DOCS "Install Doxygen-generated documentation" OFF)
 # Build OpenAL sound support
 option(OPENAL_SOUND "Build OpenAL sound support" ON)
 
+# Build Colobot with static libraries
+option(BUILD_STATIC "The default linking option for external libraries" OFF)
+
+# CBot can also be a static library
+option(CBOT_STATIC "Build CBot as static libary" ${BUILD_STATIC})
```

Comment:
Have you checked how this behaves when changing options for an existing build dir without clearing the CMake cache? I mean something like this:
```
mkdir build; cd build
cmake -DBUILD_STATIC=ON ..  # enable static builds
cmake -DBUILD_STATIC=OFF .. # disable static builds, without clearing the build dir. Are options like CBOT_STATIC stuck on "true" after this?
```

Reply:
No, I haven't, but I doubt it will change the values. It's what I would expect honestly. In my experience changing variables without clearing cache is usually a bad idea anyway and ends up badly. So I haven't even considered this use case and I'm not sure what would be the expected behavior even be. What if the flags are not set consistently? For example, some libraries are STATIC and some are not? What should switching the BUILD_STATIC variable then do? Force rewrite of the cache variables? This would overwrite any changes in the cache set by the user for example from the GUI.

Anyway, maybe we should just drop the (probably bad?) idea of mixing static and dynamic libraries (except the ones we build, like CBOT) altogether and just force everything to be build statically or dynamically with only this one flag, with maybe a changed name to USE_STATIC_EXTERNAL_LIBS or something. I'm definitely not an expert on the matter of linking though so I don't know what are good practices here.

Reply:
I'm not sure what to do here either, but I don't think I've ever seen an option to set static linking on per-library basis - it's always either all dynamic or all static.

I also prefer to clean the build dir when switching cmake options, but technically this is a use case CMake supports. One place where it could cause trouble is if somebody tries to (god forbid) use CMake GUI to configure the build options, as in that case BUILD_STATIC would do absolutely nothing.

Reply:
Also see "What is the CMake cache?" in https://cmake.org/runningcmake/ - it seems that editing the options without clearing is the recommended workflow

Reply:
After thinking about this, the best option I can think of is something like
```cmake
option(STATIC default OFF)
if (STATIC)
    option(STATIC_CBOT default ON)
    # ...
endif()
```
This still gives you the fine-grained control while keeping the simple "all static" option, and works well with the CMake cache.

Build all static: `-DSTATIC=ON`
Build all static except GLEW: `-DSTATIC=ON -DSTATIC_GLEW=OFF`
Build only GLEW static: this one is problematic, as you need to list all libraries with STATIC_*=OFF ... but it is possible, and I don't expect "single static library" to be used too often

Reply:
I think this flag should just be removed or left as it is -- as a convenient way to force everything to be static on the first run, although I'd rather remove it since it doesn't play well with the cache and I don't think it's worth going out of our way to implement a logic handling this variable how we want. It was supposed to be mainly a convenience thing.  

Also, I introduced a USE_STATIC_RUNTIME flag mainly useful for Windows (although it should be possible with GCC too, but linking statically on Linux doesn't make much sense as far as I know). It better describes what MSVC_STATIC did.

The biggest problem right now is finding the transitive dependencies, especially on Windows since we have to either link those dependencies if the build is static or install the DLLs to the output folder if the build is dynamic. We use a workaround for the former (separately for MXE and MSVC) and used a script for the latter in the past. I hope there's a better way to do this with modern CMake, but we'll have to find those dependencies ourselves since we still can't use find_package(... CONFIG) reliably. But that's a complicated issue I don't want to solve in this already huge PR.